src/OpenTK.Windowing.GraphicsLibraryFramework/JoystickState.cs

```diff
@@ -0,0 +1,197 @@
+//
+// JoystickState.cs
+//
+// Copyright (C) OpenTK
+//
+// This software may be modified and distributed under the terms
+// of the MIT license. See the LICENSE file for details.
+
+using System;
+using System.Runtime.CompilerServices;
+using System.Text;
+using OpenTK.Windowing.Common.Input;
+
+[assembly: InternalsVisibleTo("OpenTK.Windowing.Desktop")]
+
+namespace OpenTK.Windowing.GraphicsLibraryFramework
+{
+    /// <inheritdoc />
+    internal sealed class JoystickState : IJoystickState
+    {
+        private const int StatePositionCount = 2;
+
+        private int _current = StatePositionCount - 1;
+        private int _previous = StatePositionCount - 2;
+
+        private Hat[][] _hats;
+        private float[][] _axes;
+        private byte[][] _buttons;
```

Comment:
This class should only ever store 2 states. Personally I would like to see the thing (class?) that you get the input data from be different than the data structure that holds a state.
I'm not sure I have any arguments for why that is better though.

Reply:
So the question then becomes whether to pepper the magic number 2 around the code...

Storing it in a separate structure means we have to Array.Copy the values across from current to previous on every frame. Or we could put the data into a tightly packed struct and blit it, which I don't think would be as maintainable.

Reply:
Just swap the two buffers on update:
float[] tmp = last;
last = current;
current = tmp;
fillWithGlfwInputData(current);

Reply:
@varon I've made that change - IMO the arrays were cleaner

Reply:
Right, perhaps I wasn't super clear here!
The idea is to simplify by removing the index values and 2D arrays entirely.
```cs
        private Hat[] _hats;
        private Hat[] _lastHats;
        private float[] _axes;
        private float[] _lastAxes;
        private byte[] _buttons;
        private byte[] _lastButtons;
```

and then do the magical swap + copy algorithm on update.

Reply:
This makes it super obvious what's going on without needing to understand the indirect indexing or keep a mental image of what the state of those indices are.

Reply:
Do me a favour and make sure you're looking at the latest changeset. As before, this piece of conversation days "Outdated" above the code snippet.